kafkaesk/retry.py

```diff
@@ -0,0 +1,354 @@
+    timestamp: int
+    timestamp_type: int
+    key: Optional[str] = None
+    value: str
+    checksum: Optional[int] = None
+    serialized_key_size: int
+    serialized_value_size: int
+    headers: tuple
+
+    @classmethod
+    def from_consumer_record(cls, record: ConsumerRecord) -> "Record":
+        return cls(
+            topic=record.topic,
+            partition=record.partition,
+            offset=record.offset,
+            timestamp=record.timestamp,  # type: ignore
+            timestamp_type=record.timestamp_type,  # type: ignore
+            key=record.key,  # type: ignore
+            value=record.value.decode("utf-8"),  # type: ignore
+            checksum=record.checksum,  # type: ignore
+            serialized_key_size=record.serialized_key_size,  # type: ignore
+            serialized_value_size=record.serialized_value_size,  # type: ignore
+            headers=record.headers,
+        )
+
+    def to_consumer_record(self) -> ConsumerRecord:
+        # We need to convert the value back into bytes before giving this back to the consumer
+        data = self.dict()
+        data["value"] = data["value"].encode("utf-8")
+
+        return ConsumerRecord(**data)  # type: ignore
+
+
+class FailureInfo(BaseModel):
+    exception: str
+    handler_key: str
+    timestamp: datetime
+
+
+class RetryHistory(BaseModel):
+    failures: List[FailureInfo] = []
+
+
+class RetryMessage(BaseModel):
+    original_record: Record
+    retry_history: RetryHistory
+
+
+class RetryPolicy:
+    logger = logging.getLogger("kafkaesk.retry.retry_policy")
+
+    def __init__(
+        self,
+        app: "Application",
+        subscription: "Subscription",
+        retry_handlers: Optional[Dict[Type[Exception], "RetryHandler"]] = None,
+    ):
+        self.app = app
+        self.subscription = subscription
+
+        self._default_handler: "RetryHandler" = Raise()
+
+        self._handlers = retry_handlers or {}
+        self._handler_cache: Dict[Type[Exception], Tuple[str, RetryHandler]] = {}
+
+        self._ready = False
+
+        if "RetryMessage:1" not in self.app.schemas:
+            self.app.schema("RetryMessage", version=1)(RetryMessage)
+
+    async def add_retry_handler(self, exception: Type[Exception], handler: "RetryHandler") -> None:
+        if exception in self._handlers:
+            raise ValueError(f"{exception} retry handler is already set")
+
+        self._handlers[exception] = handler
+        self.logger.debug(
+            f"Retry policy added new handler {handler.__class__.__name__} for {exception}"
+        )
+
+        if self._ready is True:
+            await self._handlers[exception].initialize(self)
+            self.logger.debug(
+                f"Retry policy initialized retry handler {handler.__class__.__name__}"
+            )
+
+        # Clear handler cache when handler is added
+        self._handler_cache = {}
+
+    async def remove_retry_handler(self, exception: Type[Exception]) -> None:
```

Comment:
Do you think the `add_retry_handler` and `remove_retry_handler` are necessary APIs? Maybe the only interface we expose is from the constructor?

Reply:
You are right, I was over-designing the API for use cases we don't actually have.